### ISSUE TITLE
Using a valid ApolloLink with renderToString executes queries on the server side

### ISSUE THREAD

**Issue by peeter-tomberg** [apollographql/apollo-client#6681]
Hello! 

When using code splitting or styles components (CSS in JS), you have to iterate over the dom tree to collect chunks / extract styles.

When setting up an application with a valid apollo link and simply calling `renderToString`, Apollo Link starts executing queries in the background while the result of these queries are thrown away.

I fully understand why this is happening, `renderToString` goes through your entire application tree and renders components, including `Query` components. The problem is this is not the expectation I had. I made the assumption, that when using `getDataFromTree` we would only execute queries and I can call `renderToString` as many times as I want without any side effects.

Can we improve this with some sort of an error message or actually disable apollo link executing queries when not called in the context of `getDataFromTree` ?

**Comment by benjamn** [comment]
Definitely willing to consider adding some warnings. Can you give a concrete example of a query that's getting executed even though it's not part of a component that you're rendering? Are you using `ApolloLink` separately from `ApolloClient`, or something like that?

**Comment by peeter-tomberg** [comment]
I'll add some pseudo code to illustrate my issue, if it is still unclear I will add a concrete example tomorrow.

```
const client = new ApolloClient({
  ssrMode: true,
  link: createHttpLink({
    uri: 'http://localhost:3010',
  }),
  cache: new InMemoryCache(),
});

const GET_DOGS = gql`
  {
    dogs {
      id
      breed
    }
  }
`;

const App = (
  <ApolloProvider client={client}>
    <React.Fragment>
      <Query query={GET_DOGS}>
        {({ loading, error, data }) => {
          if (loading) return "Loading...";
          if (error) return `Error! ${error.message}`;

          return (
            <select name="dog" onChange={onDogSelected}>
              {data.dogs.map(dog => (
                <option key={dog.id} value={dog.breed}>
                  {dog.breed}
                </option>
              ))}
            </select>
          );
        }}
      </Query>
    </React.Fragment>
  </ApolloProvider>
);
```

Calling `renderToString(App);` will have the following happen:

1. Query component asks for GET_DOGS query from apollo client
2. Apollo client checks cache and does not find data
3. Apollo client executes query against apollo link
4. Apollo link executes query against http://localhost:3010

This is all happening while `renderToString` is synchronous and does not actually wait for these queries to be fetched, so they are effectively thrown away.  

**I'll be straight-forward, this is the behaviour we saw, but we were using an older version of apollo with a custom query HOC**, but this behaviour makes complete sense to me (now that I know about it). If it is confusing or you think the current active version of Apollo has addressed this concern in any way, please do let me know and I'll make a POC tomorrow.

**Comment by benjamn** [comment]
Ok, I think I see what you mean now. Thanks for those details.

In the React Apollo code, we use the React context to determine if `getDataFromTree` is active, by [checking whether `context.renderPromises` is defined](https://github.com/apollographql/apollo-client/blob/7989d5e98bd8b07c60c7f731abcc34e8636ef1fc/src/react/data/QueryData.ts#L127-L129). In Apollo Client itself, I'm not sure it's appropriate/possible to be checking the React context before making network requests. And we do want to make network requests during SSR usually, so disabling them just because you're not using `getDataFromTree` seems risky, because if we require our implementation of `getDataFromTree` to be running, that might prevent other folks from implementing async SSR utilities—but maybe we can come up with a convention that other tools could follow. I'll have to think about this some more.

That said, it sounds like you're looking for a way to prevent this problem in an existing version of Apollo Client, so future improvements might be fun to think about, but not immediately helpful.

As a workaround, when you pass `ssr: true` to `new ApolloClient({ ... })`, would it also be possible to set a default `FetchPolicy` that avoids the network?
```ts
const client = new ApolloClient({
  ssrMode: true,
  link: createHttpLink({
    uri: 'http://localhost:3010',
  }),
  cache: new InMemoryCache(),
  defaultOptions: {
    watchQuery: {
      fetchPolicy: "cache-only",
    },
    query: {
      fetchPolicy: "cache-only",
    },
});
```
Or perhaps substitute a link that doesn't do any network requests?
```ts
const client = new ApolloClient({
  ssrMode: true,
  link: new ApolloLink(operation => {
    return new Observable(observer => {
      observer.next(someFakeData);
      observer.complete();
    });
  }),
  cache: new InMemoryCache(),
});
```

**Comment by peeter-tomberg** [comment]
Hey!

I'm just wondering that perhaps explicitly stating this might be useful in docs. It came to me as a surprise that my node server is executing queries and putting a load on all other services due to this. What confused me were docs like:

https://loadable-components.com/docs/server-side-rendering/#3-setup-chunkextractor-server-side
https://styled-components.com/docs/advanced#example

That are telling me to use `extractor.collectChunks(<YourApp />)` and `renderToString(sheet.collectStyles(<YourApp />))`

So when I was first implementing complex features like code splitting and css-in-js with server side rendering what I did was literally what these guides stated.

```
extractor.collectChunks(<YourApp />);
const notSSredHtml = renderToString(sheet.collectStyles(<YourApp />));
const html = await renderToStringWithData(<YourApp />);
```

This caused performance issues during DDoS attacks, so we started checking if CPUs are under 70%  only in that case start calling `renderToStringWithData`. 

```
const shouldDoServerSideRendering = loadAverageForFiveMinutes < 0.7 * cpus().length;
const html = isServerSideRenderingEnabled ? await renderToStringWithData(App) : renderToString(App);
```

Now going through these steps, unless I have deeper knowledge of how all of this I would not expect my service to be executing graphql queries. I'm conditionally calling a function named `renderToStringWithData`. This code seems pretty valid to me, but it's actually doing quite a bit of harm. 

What we are currently doing in production is executing queries manually on the server side, and writing the query and the result to apolloCache using `writeQuery`. Then we are using this apollo link `new ApolloLink(() => null)` to make sure we are not executing queries. And only then are we doing `renderToString`.
